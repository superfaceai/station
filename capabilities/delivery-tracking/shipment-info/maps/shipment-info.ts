/* eslint-disable jest/valid-title */
/* eslint-disable jest/no-export */

import { SuperfaceTest } from '@superfaceai/testing-lib';

export const shipmentInfoTest = (
  provider: string,
  testVectors: {
    input: { trackingNumber: string; carrier?: string };
    describe: string;
    expectedResult: string;
  }[]
): void => {
  describe(`delivery-tracking/shipment-info/${provider}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = new SuperfaceTest();
    });

    describe('ShipmentInfo', () => {
      testVectors.forEach(testVector => {
        describe(testVector.describe, () => {
          it(testVector.expectedResult, async () => {
            await expect(
              superface.run({
                profile: 'delivery-tracking/shipment-info',
                provider,
                useCase: 'ShipmentInfo',
                input: testVector.input,
              })
            ).resolves.toMatchSnapshot();
          });
        });
      });
    });
  });
};
