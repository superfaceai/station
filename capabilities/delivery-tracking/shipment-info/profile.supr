"""
Shipment information
Track your shipment. Get the latest information on your shipment status. 
"""

name = "delivery-tracking/shipment-info"
version = "1.1.0"

"""
Retrieve Shipment Status
Get the current shipment status.
"""
usecase ShipmentInfo safe {
  input {
    """
    Shipment tracking number
    The identifier of shipment.
    """
    trackingNumber! string!

    """
    Carrier
    The shipment carrier identification to narrow down the results.
    """
    carrier string!
  }

  result [{
    """
    Carrier
    The name of the carrier responsible for delivery.
    """
    carrier! string

    """
    Status
    The latest shipment event.
    """
    status! ShipmentEvent

    """
    Origin
    A postal address with the origin of the shipment.
    """
    origin! Location

    """
    Shipment tracking number
    The identifier of shipment.
    """
    trackingNumber! string
    
    """
    Destination
    A postal shipping address.
    """
    destination! Location

    """
    Events
    A list of delivery tracking events.
    """
    events! [ShipmentEvent!]!

    """
    Estimated delivery date
    Estimated date and time of delivery.
    """
    estimatedDeliveryDate! string
  }]

  error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title!

    """
    Detail
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail
  }
}

"""
Shipment status
The status of the shipment. Harmonized across different carriers.
"""
model ShipmentStatus enum {
  """
  Pre-Transit
  Carrier has received information about the package but it has not yet been scanned and picked up.
  """
  pre_transit

  """
  Transit
  A package is traveling to its destination.
  """
  transit

  """
  Delivered
  The package has been delivered.
  """
  delivered

  """
  Failure
  The package encountered an error during transit.
  """
  failure

  """
  Unknown
  The shipment status is unknown.
  """
  unknown
}

"""
Shipment event
The shipment event information.
"""
model ShipmentEvent {
  """
  Timestamp
  The date and time in ISO 8601 format of the event.
  """
  timestamp! string

  """
  Shipment status
  The shipment status of the event.
  """
  statusCode! ShipmentStatus

  """
  Status
  A description of the current shipment status.
  """
  statusText! string

  """
  Location
  The location of the shipment.
  """
  location Location
}

"""
Location
A shipment location.
"""
model Location {
  address! {
    """
    Country code
    A short text string code (ISO 3166-1 alpha-2 country code) specifying the country.
    """
    countryCode string

    """
    Postal code
    Text specifying the postal code for an address.
    """
    postalCode string

    """
    Address locality
    Text specifying the name of the locality, for example a city.
    """
    addressLocality string

    """
    Street address
    The street address expressed as free form text. The street address is printed on paper as the first lines below the name.
    """
    streetAddress string
  }
}
