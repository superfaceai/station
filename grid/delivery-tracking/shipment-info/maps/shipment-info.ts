/* eslint-disable jest/valid-title */
/* eslint-disable jest/no-export */

import { SuperfaceTest } from '@superfaceai/testing';
import { nockConfig } from '../../../test-config';

export const shipmentInfoTest = (
  provider: string,
  testVectors: {
    input: { trackingNumber: string; carrier?: string };
    describe: string;
    expectedResult: string;
  }[]
): void => {
  describe(`delivery-tracking/shipment-info/${provider}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = new SuperfaceTest(
        {
          profile: 'delivery-tracking/shipment-info',
          provider,
        },
        nockConfig
      );
    });

    describe('ShipmentInfo', () => {
      testVectors.forEach(testVector => {
        describe(testVector.describe, () => {
          it(testVector.expectedResult, async () => {
            await expect(
              superface.run({
                useCase: 'ShipmentInfo',
                input: testVector.input,
              })
            ).resolves.toMatchSnapshot();
          });
        });
      });
    });
  });
};
