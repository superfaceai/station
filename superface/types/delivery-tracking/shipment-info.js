"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deliveryTrackingShipmentInfo = void 0;
var one_sdk_1 = require("@superfaceai/one-sdk");
var profile = {
    /**
     * Retrieve Shipment Status
     * Get the current shipment status.
     **/
    "ShipmentInfo": one_sdk_1.typeHelper()
};
exports.deliveryTrackingShipmentInfo = {
    "delivery-tracking/shipment-info": profile
};
