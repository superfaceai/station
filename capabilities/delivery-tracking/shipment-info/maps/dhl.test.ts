import { shipmentInfoTest } from './shipment-info';

shipmentInfoTest('dhl', [
  {
    input: {
      trackingNumber: '00340434292135100148',
    },
    describe: 'when all inputs are correct',
    expectedResult: 'should return shipment info as result',
  },
  {
    input: {
      trackingNumber: 'NOT_EXISTING_TRACKING_NUMBER',
    },
    describe: 'when inputs are invalid',
    expectedResult: 'should throw exception',
  },
]);
