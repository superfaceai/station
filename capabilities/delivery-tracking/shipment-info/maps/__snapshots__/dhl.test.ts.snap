// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delivery-tracking/shipment-info/dhl ShipmentInfo when all inputs are correct should return shipment info as result 1`] = `
Object {
  "value": Array [
    Object {
      "carrier": "parcel-de",
      "destination": Object {
        "address": Object {
          "countryCode": "DE",
        },
      },
      "estimatedDeliveryDate": null,
      "events": Array [
        Object {
          "location": undefined,
          "statusCode": "pre_transit",
          "statusText": "The instruction data for this shipment have been provided by the sender to DHL electronically",
          "timestamp": "2021-07-06T19:35:00",
        },
      ],
      "origin": Object {
        "address": Object {
          "countryCode": "DE",
        },
      },
      "status": Object {
        "location": undefined,
        "statusCode": "pre_transit",
        "statusText": "The instruction data for this shipment have been provided by the sender to DHL electronically",
        "timestamp": "2021-07-06T19:35:00",
      },
      "trackingNumber": "00340434292135100148",
    },
  ],
}
`;

exports[`delivery-tracking/shipment-info/dhl ShipmentInfo when inputs are invalid should throw exception 1`] = `
Object {
  "error": "HTTPError: Expected HTTP error
AST Path: definitions[0].statements[1]
Original Map Location: Line 11, column 3",
}
`;
